lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -459,46 +457,50 @@ def access_token_url
       end
 
       def setup_access_token
-        request = 'grant_type=client_credentials'
-        begin
-          raw_response = ssl_post(access_token_url, request, access_token_header)
-        rescue ResponseError => e
-          raise OAuthResponseError.new(e)
-        else
-          response = parse(raw_response)
+        return if @options[:secret_key] || @options[:access_token]
 
-          if (access_token = response['access_token'])
-            access_token
-          else
-            raise OAuthResponseError.new(response)
-          end
-        end
+        response = parse(ssl_post(access_token_url, 'grant_type=client_credentials', access_token_header))
+        @options[:access_token] = response['access_token']
+      rescue ResponseError => e
+        @multiresponse = MultiResponse.new
+        @multiresponse.process { Response.new(false, e.response.body.to_s, {}, test: test?) }
       end
 
       def commit(action, post, options, authorization = nil, method = :post)
+        setup_access_token
         begin
           raw_response = ssl_request(method, url(action, authorization), post.nil? || post.empty? ? nil : post.to_json, headers(action, options))
           response = parse(raw_response)
           response['id'] = response['_links']['payment']['href'].split('/')[-1] if action == :capture && response.key?('_links')
-          source_id = authorization if action == :unstore
         rescue ResponseError => e
+          if e.response.to_s.match?(/Unauthorized/) && @options[:access_token] && action != :tokens
+            @options[:access_token] = nil
+            return commit(action, post, options, authorization, method)
```

Comment:
I'm a bit worried this could cause a constant loop if a transaction fails and it going through commit again and again.

Reply:
I also don't think Unauthorized is a good error message to retry since I think this is a clear indication that the credentials are incorrect. The error that brought this ticket is Failed with 429 Too Many Requests so I'm not sure if this is a better one

Reply:
I changed the implementation of this part, to do this I added a new field in the gateway options called `expires`, now we prevent the loop and we know better what the error is in case it occurs